[docs] Clarify withContiguousStorageIfAvailable documentation

glessard opened the pull request:
Improves documentation of the `withContiguousStorageIfAvailable` and `withContiguousMutableStorageIfAvailable` functions. This does not change the requirements, but states them more forcefully.

glessard committed:
[stdlib] add a missing deprecation

glessard commented:
@swift-ci please smoke test

lorentey commented on stdlib/public/core/MutableCollection.swift:
```diff
@@ -173,33 +173,34 @@ where SubSequence: MutableCollection
-  /// Often, the optimizer can eliminate bounds- and uniqueness-checks
-  /// within an algorithm, but when that fails, invoking the
-  /// same algorithm on `body`\ 's argument lets you trade safety for
-  /// speed.
```
I think we should keep these documented, even if they are deprecated

glessard replied:
I restored the documentation, even though they're both deprecated and internal.

lorentey commented on stdlib/public/core/Sequence.swift:
```diff
@@ -384,18 +384,27 @@ public protocol Sequence {
-  /// slice's `startIndex`.
+  /// must guarantee that contiguous storage to any valid `SubSequence`
+  /// can be generated by slicing `buffer` with a range formed from the
+  /// distances to the subsequence's `startIndex` and `endIndex`, respectively.
```
This passage is a definite improvement, but I find it a bit difficult to understand what it means. Is it equivalent to stating that the order of elements in the buffer must match the order they appear in the collection?

glessard replied:
I think it does. That would be a good way to say it.

lorentey reviewed (approved):
Looks great! 👍 A couple of comments below

glessard committed:
[stdlib] improve documentation of `withContiguous[Mutable]StorageIfAvailable`

- clarify the requirement that the entire collection must be accessible
- clarify the requirement surrounding subsequences / slices
- add parameter descriptions
- specify that buffer cannot be replaced

glessard commented:
@swift-ci please smoke test

glessard commented:
@swift-ci please smoke test

glessard commented:
@swift-ci please smoke test and merge

glessard committed:
[gardening] fix copy-pasta

glessard commented:
@swift-ci please smoke test and merge

glessard commented:
@swift-ci please test macOS platform

glessard commented:
@swift-ci please smoke test linux platform

glessard commented:
@swift-ci please test macOS platform

glessard commented:
The Linux errors are in swift-docc tests.

glessard commented:
@swift-ci smoke test and merge

glessard commented:
@swift-ci please smoke test Linux platform
